Lib/functools.py

```diff
@@ -949,6 +952,12 @@ def _method(*args, **kwargs):
         _method.__isabstractmethod__ = self.__isabstractmethod__
         _method.register = self.register
         update_wrapper(_method, self.func)
+
+        try:
+            obj.__dict__[self.attrname] = _method
+        except AttributeError:
+            pass  # not all objects have __dict__ (e.g. classes with __slots__)
```

Comment:
Have you done benchmarks for the impact this has on slotted classes? I worry that this could make `@singledispatchmethod` a fair bit slower for slotted classes, as we'll hit the `except AttributeError` branch every time.

Raising-then-catching an exception is pretty slow with CPython -- "zero-cost exception handling" in Python 3.11 sped exception handling up quite a lot for the case where an exception is never raised, but actually made exception handling even slower for the raise-then-catch case.

A `hasattr()` check, rather than a `try`/`except` check, _should_ be less costly for slotted classes, while not slowing things down at all for classes without `__slots__` (since for classes without `__slots__`, we'll only get to this point in the function the first time that `__get__` is called). (I haven't checked whether my assumption about performance there is correct, however; you might want to.)

```suggestion
        if hasattr(obj, "__dict__"):
            obj.__dict__[self.attrname] = _method
```

Reply:
@AlexWaygood The `hasattr` approach is indeed much faster than the `try-except`. I updated the code, but can only run benchmarks a little later.